appendix-basic65.tex

```diff
@@ -3470,8 +3569,10 @@ \subsection{EDMA}
 \begin{description}[leftmargin=2cm,style=nextline]
 \item [Token:] \$FE \$21
 \item [Format:] {\bf EDMA} command{\bf,} length{\bf,} source{\bf,}
-                 target [{\bf,} sub{\bf,} mod]
+                 target
```

Comment:
Just wanted the backstory on the removal/hiding of those extra params bit shifter added: "target **[{\bf,} sub{\bf,} mod]**"
- have they genuinely been removed from the rom?
- or are we just hiding them due to confusion described in past discord convos, in relation to his ticket/pr?
  - https://github.com/MEGA65/mega65-user-guide/issues/518
  - https://github.com/MEGA65/mega65-user-guide/pull/517

Reply:
These arguments don't do anything yet due to gaps in the DMA. They set records $09 and $0a+$0b at the end of the 12-byte EDMA structure, but these are non-functional in the DMA right now. Of the addressing modes ($09), only Linear (the defaut) and Hold are implemented, and IIRC setting Hold via $09 doesn't work yet. Even if Hold did work via $09, you can set Hold using the upper bits of the address BANK byte. (Notably, Direction is only available on the BANK byte and is not an addressing mode. As noted in 518, you can hack this by setting those bits in the address field in the EDMA command.)

IMO, this is the wrong API for a BASIC command to begin with. I would much rather EDMA support explicit flag arguments, as a separate bit field or a string with an internal syntax. Because the current arguments do nothing, now is the time to rescind them, so we can replace them with better arguments later. Even if we do end up liking (or settling for) this API, we can reinstate the arguments when the fields are fixed in the core.

I'd rather that EDMA do something like this:
```
EDMA command, length, source, target, source_mode, target_mode, modulo_size, job_options
```
A mode argument can be a number to set flags numerically: use bits 0 and 1 to select the four addressing modes (linear, modulo, hold, XY MOD), and bits 6 and 7 to set direction and I/O register visibility. Or it could be a string value, setting those flags by letter, e.g. "LR" for linear reversed, or "M" for modulo. (I wouldn't offer bare letters in this case, so flags could be from a variable.)

I'm not sure what's most useful for exposing job options to BASIC. Maybe it's a string with special PETSCII-based syntax, or a variable argument list that just accepts option codes and parameter bytes in alternating positions. Maybe we never expose job options to BASIC.

Reply:
No probs, just happy to capture the backstory of thought on this.